lightning/src/ln/channelmanager.rs

```diff
@@ -7221,6 +7212,27 @@ where
 			);
 		}
 		let counterparty_node_id = prev_hop.counterparty_node_id.expect("Checked immediately above");
+		let mut peer_state = peer_state_opt.expect("peer_state_opt is always Some when the counterparty_node_id is Some");
+
+		let mut preimage_update = ChannelMonitorUpdate {
+			update_id: 0, // set in set_closed_chan_next_monitor_update_id
+			counterparty_node_id: prev_hop.counterparty_node_id,
+			updates: vec![ChannelMonitorUpdateStep::PaymentPreimage {
+				payment_preimage,
+				payment_info,
+			}],
+			channel_id: Some(prev_hop.channel_id),
+		};
+
+		// Note that the below is race-y - we set the `update_id` here and then drop the peer_state
+		// lock before applying the update in `apply_post_close_monitor_update` (or via the
+		// background events pipeline). During that time, some other update could be created and
+		// then applied, resultin in `ChannelMonitorUpdate`s being applied out of order and causing
+		// a panic.
+		Self::set_closed_chan_next_monitor_update_id(&mut *peer_state, prev_hop.channel_id, &mut preimage_update);
```

Comment:
Looks like this is the only place where `set_closed_chan_next_monitor_update_id ` is called. Maybe we should move the code here as it seems some checks on the `ChannelMonitorUpdate` (i.e., `is_closing_unupdated_monitor`) aren't necessary?